cylc/flow/scripts/reinstall.py

```diff
@@ -0,0 +1,171 @@
+from cylc.flow.option_parsers import CylcOptionParser as COP
+from cylc.flow.pathutil import get_workflow_run_dir
+from cylc.flow.platforms import get_platform
+from cylc.flow.suite_files import (
+    get_workflow_source_dir,
+    reinstall_workflow,
+    SuiteFiles)
+from cylc.flow.terminal import cli_function
+
+
+def get_option_parser():
+    parser = COP(
+        __doc__, comms=True, prep=True,
+        argdoc=[("[NAMED_RUN]", "Named run. e.g. my-flow/run1")
+                ])
+
+    # If cylc-rose plugin is available ad the --option/-O config
+    try:
+        __import__('cylc.rose')
+        parser.add_option(
+            "--opt-conf-key", "-O",
+            help=(
+                "Use optional Rose Config Setting"
+                "(If Cylc-Rose is installed)"
+            ),
+            action="append",
+            default=[],
+            dest="opt_conf_keys"
+        )
+        parser.add_option(
+            "--define", '-D',
+            help=(
+                "Each of these overrides the `[SECTION]KEY` setting in a "
+                "`rose-suite.conf` file. "
+                "Can be used to disable a setting using the syntax "
+                "`--define=[SECTION]!KEY` or even `--define=[!SECTION]`."
+            ),
+            action="append",
+            default=[],
+            dest="defines"
+        )
+        parser.add_option(
+            "--define-suite", "--define-flow", '-S',
+            help=(
+                "As `--define`, but with an implicit `[SECTION]` for "
+                "workflow variables."
+            ),
+            action="append",
+            default=[],
+            dest="define_suites"
+        )
+    except ImportError:
+        pass
+
+    return parser
+
+
+@cli_function(get_option_parser)
+def main(parser, opts, named_run=None):
+    if not named_run:
+        source = get_workflow_source_dir(Path.cwd())
+        if source is None:
+            raise WorkflowFilesError("Cannot find an installed flow in $PWD. "
+                                     "Try again, stating the named run. "
+                                     "E.g. cylc reinstall my-flow/run1")
+        base_run_dir = Path(
+            get_platform()['run directory'].replace('$HOME', '~')).expanduser()
+        named_run = Path.cwd().relative_to(Path(base_run_dir).resolve())
+    run_dir = Path(get_workflow_run_dir(named_run)).expanduser()
+    if not run_dir.exists():
+        raise(WorkflowFilesError(f"\"{run_dir}\" does not exist. "
+                                 f"Check path provided: \"{named_run}\""))
+    if run_dir.name in [SuiteFiles.FLOW_FILE, SuiteFiles.SUITE_RC]:
+        run_dir = run_dir.parent
+        named_run = named_run.rsplit('/', 1)[0]
+    source = get_workflow_source_dir(run_dir)
+    if not source:
+        raise(WorkflowFilesError("Could not establish source directory. "
+                                 f"Check path provided: \"{named_run}\"")
```

Comment:
This error message could by a little cryptic, I think the main use cases are when the user has mistyped the flow reg or is trying to reinstall somewhere it wasn't installed in the first place.